discourse_check_cleanup.py

```diff
@@ -0,0 +1,306 @@
+#!/usr/bin/env python
+
+# # Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Cleanup utility after running the action, primarily for testing purposes."""
+
+import argparse
+import contextlib
+import json
+import logging
+import sys
+
+from src.discourse import Discourse, create_discourse
+from src.exceptions import DiscourseError
+from src.types_ import ActionResult
+
+
+def main():
+    """Clean up created Discourse pages."""
+    logging.basicConfig(level=logging.INFO)
+
+    parser = argparse.ArgumentParser(
+        prog="DiscourseCleanup",
+        description="Delete posts created on discourse during an action execution.",
+    )
+    parser.add_argument("urls_with_actions", help="The pages that were created during execution")
+    parser.add_argument(
+        "discourse_config", help="The discourse configuration used to create the pages"
+    )
+    parser.add_argument(
+        "--check-draft", help="Check that the draft test succeeded", action="store_true"
+    )
+    parser.add_argument(
+        "--check-create", help="Check that the create test succeeded", action="store_true"
+    )
+    parser.add_argument(
+        "--check-delete-topics",
+        help="Check that the delete_topics test succeeded",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--check-delete", help="Check that the delete test succeeded", action="store_true"
+    )
+    parser.add_argument(
+        "--check-only", help="Skip cleanup and only run any checks", action="store_true"
+    )
+    args = parser.parse_args()
+    urls_with_actions = json.loads(args.urls_with_actions)
+    discourse_config = json.loads(args.discourse_config)
+
+    discourse = create_discourse(**discourse_config)
+
+    check_result = True
+    if args.check_draft:
+        check_result = check_draft(urls_with_actions=urls_with_actions)
+    if args.check_create:
+        check_result = check_create(urls_with_actions=urls_with_actions, discourse=discourse)
+    if args.check_delete_topics:
+        check_result = check_delete_topics(
+            urls_with_actions=urls_with_actions, discourse=discourse
+        )
+    if args.check_delete:
+        check_result = check_delete(urls_with_actions=urls_with_actions, discourse=discourse)
+
+    if not args.check_only:
+        for url in urls_with_actions.keys():
+            with contextlib.suppress(DiscourseError):
+                discourse.delete_topic(url=url)
+
+    sys.exit(0 if check_result else 1)
+
+
+def _check_url_count(
```

Comment:
I'd expect all of this methods to throw exceptions instead of controlling "exceptional" flow with booleans

Reply:
I'll provide the reasoning for the current design and see what you think. This utility is mainly for the action integration tests and makes use of the action output to validate whether the test has passed. If any of the checks fail, it doesn't mean that the utility has encountered an error or exception since the utility doesn't depend on whether the checks pass just to execute the checks and report back whether they pass or fail, similar to how a linter might be written. In the case of a linter, an issue it finds isn't an exception for the linter, the linter is working fine, it just needs to inspect the code and report all the problems. The user of the linter can then decide whether the linter failing is exceptional flow. In this case, the GitHub workflow will use the return value of this utility in the same way it would use the 0 or 1 return value of a linter. Does that make sense?